src/function/video.rs

```diff
@@ -0,0 +1,426 @@
+                        color_matching.color_primaries.to_string(),
+                    )?;
+                    self.dir.write(
+                        frame.format.color_matching_path().join("bTransferCharacteristics"),
+                        color_matching.transfer_characteristics.to_string(),
+                    )?;
+                    self.dir.write(
+                        frame.format.color_matching_path().join("bMatrixCoefficients"),
+                        color_matching.matrix_coefficients.to_string(),
+                    )?;
+                    self.dir.symlink(&color_matching_path, frame.format.color_matching_link_path())?;
+                } else {
+                    log::warn!("Color matching information already exists for format {:?}", frame.format);
+                }
+            }
+        }
+
+        // header linking format descriptors and associated frames to header after creating
+        // otherwise cannot add new frames
+        self.dir.create_dir_all("streaming/header/h")?;
+        self.dir.create_dir_all("control/header/h")?;
+
+        for format in formats_to_link {
+            self.dir.symlink(format.group_path(), format.header_link_path())?;
+        }
+
+        // supported speeds: all linked but selected based on gadget speed: https://github.com/torvalds/linux/blob/master/drivers/usb/gadget/function/f_uvc.c#L732
+        self.dir.symlink("streaming/header/h", "streaming/class/fs/h")?;
+        self.dir.symlink("streaming/header/h", "streaming/class/hs/h")?;
+        self.dir.symlink("streaming/header/h", "streaming/class/ss/h")?;
+        self.dir.symlink("control/header/h", "control/class/fs/h")?;
+        self.dir.symlink("control/header/h", "control/class/ss/h")?;
+
+        // controls
+        if let Some(processing_controls) = self.builder.processing_controls {
+            self.dir.write("control/processing/default/bmControls", processing_controls.to_string())?;
+        }
+
+        // terminal
+        if let Some(camera_controls) = self.builder.camera_controls {
+            self.dir.write("control/terminal/camera/default/bmControls", camera_controls.to_string())?;
+        }
+
+        // bandwidth configuration
+        if let Some(interval) = self.builder.streaming_interval {
+            self.dir.write("streaming_interval", interval.to_string())?;
+        }
+        if let Some(max_burst) = self.builder.streaming_max_burst {
+            self.dir.write("streaming_maxburst", max_burst.to_string())?;
+        }
+        if let Some(max_packet) = self.builder.streaming_max_packet {
+            self.dir.write("streaming_maxpacket", max_packet.to_string())?;
+        }
+
+        Ok(())
+    }
+}
+
+/// USB Video Class (UVC) function.
+#[derive(Debug)]
+pub struct Uvc {
+    dir: FunctionDir,
+}
+
+impl Uvc {
+    /// Creates a new USB Video Class (UVC) builder with f_uvc video defaults.
+    pub fn builder() -> UvcBuilder {
```

Comment:
You could add the `frames` parameter from `new` here instead, if it is mandatory for the gadget.

Reply:
I left it as plain builder since one might want the `UvcBuilder` to edit the other fields. Like the MSD, I’ve added a `with_frames` to the `UvcBuilder` so it’s easy to add `Frame`s then change other parameters like test/video.rs